lib/nodes/iec61850_goose.cpp

```diff
@@ -0,0 +1,577 @@
+
+	switch (*descriptor->mms_type) {
+	case MmsType::MMS_BOOLEAN:
+		return MmsValue_newBoolean(signal_data.b);
+	case MmsType::MMS_INTEGER:
+		return newMmsInteger(signal_data.i, meta.size);
+	case MmsType::MMS_UNSIGNED:
+		return newMmsUnsigned(static_cast<uint64_t>(signal_data.i), meta.size);
+	case MmsType::MMS_FLOAT:
+		return newMmsFloat(signal_data.f, meta.size);
+	default:
+		assert(!"unreachable");
+	}
+}
+
+MmsValue * GooseValue::newMmsInteger(int64_t i, GooseValue::MetaSize size) {
+	auto mms_integer = MmsValue_newInteger(size.value);
+
+	switch (size.value) {
+	case 8:
+		MmsValue_setInt8(mms_integer, static_cast<int8_t>(i));
+		return mms_integer;
+	case 16:
+		MmsValue_setInt16(mms_integer, static_cast<int16_t>(i));
+		return mms_integer;
+	case 32:
+		MmsValue_setInt32(mms_integer, static_cast<int32_t>(i));
+		return mms_integer;
+	case 64:
+		MmsValue_setInt64(mms_integer, static_cast<int64_t>(i));
+		return mms_integer;
+	default:
+		assert(!"unreachable");
+	}
+}
+
+MmsValue * GooseValue::newMmsUnsigned(uint64_t u, GooseValue::MetaSize size) {
+	auto mms_unsigned = MmsValue_newUnsigned(size.value);
+
+	switch (size.value) {
+	case 8:
+		MmsValue_setUint8(mms_unsigned, static_cast<uint8_t>(u));
+		return mms_unsigned;
+	case 16:
+		MmsValue_setUint16(mms_unsigned, static_cast<uint16_t>(u));
+		return mms_unsigned;
+	case 32:
+		MmsValue_setUint32(mms_unsigned, static_cast<uint32_t>(u));
+		return mms_unsigned;
+	default:
+		assert(!"unreachable");
+	}
+}
+
+MmsValue * GooseValue::newMmsFloat(double d, GooseValue::MetaSize size) {
+	switch (size.value) {
+	case 32:
+		return MmsValue_newFloat(static_cast<float>(d));
+	case 64:
+		return MmsValue_newDouble(d);
+	default:
+		assert(!"unreachable");
+	}
+}
+
+std::optional<GooseValue::Descriptor const *> GooseValue::lookupMmsType(int mms_type)
+{
+	auto check = [mms_type] (Descriptor descriptor) {
+		return descriptor.mms_type == mms_type;
+	};
+
+	auto descriptor = std::find_if(begin(descriptors), end(descriptors), check);
+	if (descriptor != end(descriptors))
+		return &*descriptor;
+	else
+		return std::nullopt;
+}
+
+std::optional<GooseValue::Descriptor const *> GooseValue::lookupName(char *name)
+{
+	auto check = [name] (Descriptor descriptor) {
+		return !strcmp(descriptor.name, name);
+	};
+
+	auto descriptor = std::find_if(begin(descriptors), end(descriptors), check);
+	if (descriptor != end(descriptors))
+		return &*descriptor;
+	else
+		return std::nullopt;
+}
+
+GooseValue::GooseValue()
+: signal_data({})
+, descriptor(&descriptors[0])
+{
+}
+
+GooseValue::GooseValue(GooseValue::Descriptor const *descriptor, SignalData data)
+: signal_data(data)
+, descriptor(descriptor)
+{
+}
+
+void GooseNode::onEvent(GooseSubscriber subscriber, GooseNode::InputEventContext &ctx) noexcept
+{
+	if (!GooseSubscriber_isValid(subscriber) || GooseSubscriber_needsCommission(subscriber))
+		return;
+
+	if (	ctx.subscriber_config.trigger == InputTrigger::CHANGE
```

Comment:
Please use the same indention as in the rest of the code. And sorry for beeing so picky about it...